src/Microsoft.ML.Transforms/SvmLight/SvmLightLoader.cs

```diff
@@ -0,0 +1,851 @@
+        /// to a numeric vector of indices. This class is used in case the indices in the input file are numeric. For the case where the input file contains
+        /// feature names, instead of a <see cref="CustomMappingTransformer{TSrc, TDst}"/>, we use a <see cref="ValueToKeyMappingTransformer"/>.
+        /// </summary>
+        private sealed class Indices
+        {
+            public VBuffer<int> FeatureKeys;
+
+            public static void ParseIndicesToOneBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index > 0)
+                        editor.Values[i] = index - 1;
+                    else
+                        editor.Values[i] = -1;
+                }
+                output.FeatureKeys = editor.Commit();
+            }
+
+            public static void ParseIndicesToZeroBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index >= 0)
+                        editor.Values[i] = index;
+                    else
+                        editor.Values[i] = -1;
+                }
+                output.FeatureKeys = editor.Commit();
+            }
+        }
+
+        /// <summary>
+        /// This class and the <see cref="IntermediateOut"/> class are used by the <see cref="CustomMappingTransformer{TSrc, TDst}"/>
+        /// that maps a vector of indices and a vector of values into a single <see cref="VBuffer{T}"/> of values. When the indices
+        /// originate from the <see cref="ValueToKeyMappingTransformer"/> (in case features are specified by name), <see cref="IntermediateOutKeys"/>
+        /// is used, and when they originate from a <see cref="CustomMappingTransformer{TSrc, TDst}"/> that produces an <see cref="Indices"/>,
+        /// <see cref="IntermediateOut"/> is used.
+        /// </summary>
+        private sealed class IntermediateOutKeys
+        {
+            public VBuffer<uint> FeatureKeys;
+            public VBuffer<float> FeatureValues;
+        }
+
+        private sealed class IntermediateOut
+        {
+            public VBuffer<int> FeatureKeys;
+            public VBuffer<float> FeatureValues;
+        }
+
+        private sealed class Output
+        {
+            public VBuffer<float> Features;
+        }
```

Comment:
I think you mentioned that this is a dense array. Would it be better to keep it as a sparse array so that the algorithm that is making use of this data can decide at a later time if it needs to convert this into a dense array?

Reply:
I switched to using a `BufferBuilder` instead of the `VBufferEditor`, since it knows how to handle indices out of order. There might be perf implications though, not sure which is better to use in this case.

---
In reply to: [332210125](https://github.com/dotnet/machinelearning/pull/4190#discussion_r332210125) [](ancestors = 332210125)